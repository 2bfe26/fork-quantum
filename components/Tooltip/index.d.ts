import React from 'react';

export interface TooltipProps {
    text: string;
    placement?: 'top' | 'right' | 'bottom' | 'left';
    visible?: boolean;
    multiline?: boolean;
    children: React.ReactNode[] | React.ReactNode;
    theme?: {
        spacing?: {};
        colors?: {};
        baseFontSize?: number;
    };
}

export default class Tooltip extends React.Component<TooltipProps> {}
