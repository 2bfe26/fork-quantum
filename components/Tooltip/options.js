import Colors from '../Colors';

const arrowSize = '5px';
const upDownBorders = `
  border-left: ${arrowSize} solid transparent;
  border-right: ${arrowSize} solid transparent;
  left: 50%;
  transform: translateX(-50%);
`;

const sideBorders = `
  border-top: ${arrowSize} solid transparent;
  border-bottom: ${arrowSize} solid transparent;
  top: 50%;
  transform: translateY(-50%);
`;

const placement = {
  arrowPosition: {
    bottom: `
      ${upDownBorders}
      border-bottom: ${arrowSize} solid;
      border-bottom-color: inherit;
      top: -${arrowSize};
    `,
    top: `
      ${upDownBorders}
      border-top: ${arrowSize} solid;
      border-top-color: inherit;
      bottom: -${arrowSize};
    `,
    left: `
      ${sideBorders}
      border-left: ${arrowSize} solid;
      border-left-color: inherit;
      right: -${arrowSize};
    `,
    right: `
      ${sideBorders}
      border-right: ${arrowSize} solid;
      border-right-color: inherit;
      left: -${arrowSize};
    `,
  },
};

const skins = {
  info: Colors.SECONDARY['500'],
  danger: Colors.DANGER['500'],
  success: Colors.ACTION['400'],
  warning: Colors.WARMING['400'],
};

export {
  skins,
  placement,
};
