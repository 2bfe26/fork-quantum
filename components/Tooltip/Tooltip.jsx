import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Colors from '../Colors';
import { skins, placement } from './options';

const tipSkin = ({ skin }) => skins[skin] || skins.info;

const Tip = styled.div`
  background-color: ${tipSkin};
  border-color: ${tipSkin};
  border-radius: 2px;
  color: ${Colors.WHITE};
  font-size: 14px;
  font-weight: bold;
  opacity: ${props => (props.show ? '1' : '0')};
  padding: 5px 20px;
  position: absolute;
  text-align: center;
  transition: opacity 0.2s ease-in-out, visibility 0.2s ease-in-out;
  visibility: ${props => (props.show ? 'visible' : 'hidden')};
  z-index: 100;

  ${placement.tipPosition}

  left: ${props => props.offset}%;

  &:before {
    content: '';
    position: absolute;
    ${props => placement.arrowPosition[props.place]};
  }
`;

const Wrapper = styled.div`
  display: ${props => (props.slider ? 'block' : 'inline-block')};
  position: ${props => (props.slider ? 'initial' : 'relative')};
  cursor: default;
  white-space: nowrap;
`;

class Tooltip extends Component {
  constructor(props) {
    super(props);

    this.state = { show: false, width: null, height: null };
  }

  componentDidMount() {
    this.measure();
  }

  shouldComponentUpdate(nextProps, nextState) {
    return (
      this.props.text !== nextProps.text ||
      this.state.show !== nextState.show ||
      this.state.width !== nextState.width ||
      this.state.height !== nextState.height
    );
  }

  componentDidUpdate() {
    this.measure();
  }

  measure() {
    const {
      clientWidth,
      clientHeight,
    } = this.tip;

    this.setState({ width: clientWidth, height: clientHeight });
  }

  handleEnter = () => {
    this.setState({ show: true });
  }

  handleLeave = () => {
    this.setState({ show: false });
  }

  render() {
    const {
      children,
      skin,
      place,
      text,
      slider,
      offset,
    } = this.props;

    const {
      width,
      height,
      show,
    } = this.state;

    return (
      <Wrapper
        onMouseEnter={!slider ? this.handleEnter : () => {}}
        onMouseLeave={!slider ? this.handleLeave : () => {}}
        slider={slider}
      >
        <Tip
          skin={skin}
          innerRef={(tip) => { this.tip = tip; }}
          place={place}
          width={width}
          height={height}
          show={slider ? true : show}
          slider={slider}
          offset={offset}
        >
          {text}
        </Tip>
        {children}
      </Wrapper>
    );
  }
}

Tooltip.propTypes = {
  children: PropTypes.node.isRequired,
  skin: PropTypes.oneOf([
    'info',
    'danger',
    'success',
    'warning',
  ]),
  slider: PropTypes.bool,
  place: PropTypes.oneOf([
    'top',
    'right',
    'bottom',
    'left',
  ]),
  text: PropTypes.string,
  offset: PropTypes.number,
};

Tooltip.defaultProps = {
  skin: 'info',
  place: 'top',
  text: 'Tooltip',
  slider: false,
  offset: 0,
};

export default Tooltip;
