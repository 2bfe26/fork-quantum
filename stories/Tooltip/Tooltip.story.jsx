import React from 'react';
import { storiesOf } from '@storybook/react';
import styled from 'styled-components';
import AutoExample from '../../.storybook/components/AutoExample';
import { Tab } from '../../.storybook/components/TabbedView';
import Tooltip from '../../components/Tooltip';
import Logo from '../../components/Logo';

const TipGrid = styled.div`
  display: grid;
  grid-template-columns: repeat(4, 110px);
  margin: 0 15px;
`;

const tabExample = (
  <Tab title="Example">
    <h3>Skins</h3>
    <TipGrid>
      <div>
        <Tooltip skin="info">Info</Tooltip>
      </div>
      <div>
        <Tooltip skin="danger">Danger</Tooltip>
      </div>
      <div>
        <Tooltip skin="success">Success</Tooltip>
      </div>
      <div>
        <Tooltip skin="warning">Warning</Tooltip>
      </div>
    </TipGrid>

    <h3>Placement</h3>
    <TipGrid>
      <div>
        <Tooltip place="top">Top</Tooltip>
      </div>
      <div>
        <Tooltip place="right">Right</Tooltip>
      </div>
      <div>
        <Tooltip place="bottom">Bottom</Tooltip>
      </div>
      <div>
        <Tooltip place="left">Left</Tooltip>
      </div>
    </TipGrid>

    <h3>Others</h3>
    <TipGrid>
      <div>
        <Tooltip text="Catho Empresas">
          <Logo.Business />
        </Tooltip>
      </div>
      <div>
        <Tooltip place="right" text="Catho">
          <Logo />
        </Tooltip>
      </div>
      <div>
        <Tooltip place="bottom" text="Catho Educação">
          <Logo.Education />
        </Tooltip>
      </div>
    </TipGrid>
  </Tab>
);

storiesOf('8. Tooltip', module).add('Basic', () => (
  <AutoExample
    component={Tooltip}
    additionalTabs={tabExample}
    componentProps={{
      children: 'Hover me',
    }}
  />
));
