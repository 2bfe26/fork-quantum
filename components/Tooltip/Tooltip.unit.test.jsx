import React from 'react';
import { mount } from 'enzyme';
import Tooltip from './Tooltip';

describe('Tooltip component ', () => {
  it('Should match the snapshot', () => {
    const tooltip = mount(<Tooltip />);
    expect(tooltip.html()).toMatchSnapshot();
  });
});
