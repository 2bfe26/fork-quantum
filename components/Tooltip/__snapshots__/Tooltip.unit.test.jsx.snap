// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip component  All positions Should match the snapshot when place is bottom 1`] = `"<div class=\\"Tooltip__Wrapper-eRZUrI eXyEfk\\"><div class=\\"Tooltip__Tip-ewJBsm lgArYO\\" offset=\\"\\" width=\\"0\\" height=\\"0\\">Tooltip</div>Hover Me</div>"`;

exports[`Tooltip component  All positions Should match the snapshot when place is left 1`] = `"<div class=\\"Tooltip__Wrapper-eRZUrI eXyEfk\\"><div class=\\"Tooltip__Tip-ewJBsm dFIvSR\\" offset=\\"\\" width=\\"0\\" height=\\"0\\">Tooltip</div>Hover Me</div>"`;

exports[`Tooltip component  All positions Should match the snapshot when place is right 1`] = `"<div class=\\"Tooltip__Wrapper-eRZUrI eXyEfk\\"><div class=\\"Tooltip__Tip-ewJBsm faxpBF\\" offset=\\"\\" width=\\"0\\" height=\\"0\\">Tooltip</div>Hover Me</div>"`;

exports[`Tooltip component  All positions Should match the snapshot when place is top 1`] = `"<div class=\\"Tooltip__Wrapper-eRZUrI eXyEfk\\"><div class=\\"Tooltip__Tip-ewJBsm eZyyod\\" offset=\\"\\" width=\\"0\\" height=\\"0\\">Tooltip</div>Hover Me</div>"`;

exports[`Tooltip component  All skins Should match the snapshot skin is info 1`] = `"<div class=\\"Tooltip__Wrapper-eRZUrI eXyEfk\\"><div class=\\"Tooltip__Tip-ewJBsm eZyyod\\" offset=\\"\\" width=\\"0\\" height=\\"0\\">Tooltip</div>Hover Me</div>"`;

exports[`Tooltip component  All skins Should match the snapshot when skin is danger 1`] = `"<div class=\\"Tooltip__Wrapper-eRZUrI eXyEfk\\"><div class=\\"Tooltip__Tip-ewJBsm htRGCW\\" offset=\\"\\" width=\\"0\\" height=\\"0\\">Tooltip</div>Hover Me</div>"`;

exports[`Tooltip component  All skins Should match the snapshot when skin is success 1`] = `"<div class=\\"Tooltip__Wrapper-eRZUrI eXyEfk\\"><div class=\\"Tooltip__Tip-ewJBsm gbQtNq\\" offset=\\"\\" width=\\"0\\" height=\\"0\\">Tooltip</div>Hover Me</div>"`;

exports[`Tooltip component  All skins Should match the snapshot when skin is warning 1`] = `"<div class=\\"Tooltip__Wrapper-eRZUrI eXyEfk\\"><div class=\\"Tooltip__Tip-ewJBsm hCZIeC\\" offset=\\"\\" width=\\"0\\" height=\\"0\\">Tooltip</div>Hover Me</div>"`;
